parallel-exec/lib/lib.go

```diff
@@ -0,0 +1,418 @@
+		runnerOptions.Timer = timer
+	}
+}
+
+// Runner runs the commands.
+type Runner interface {
+	// Run the commands.
+	//
+	// Return error if there was an initialization error, or any of
+	// the running commands returned with a non-zero exit code.
+	Run(cmds []*exec.Cmd) error
+}
+
+// NewRunner returns a new Runner.
+func NewRunner(options ...RunnerOption) Runner {
+	return newRunner(options...)
+}
+
+type event struct {
+	E EventType
+	T time.Time
+	F map[string]string
+}
+
+func newEvent(e EventType, t time.Time, f map[string]string) *event {
+	return &event{e, t, f}
+}
+
+func newStartedEvent(t time.Time) *event {
+	return newEvent(EventTypeStarted, t, nil)
+}
+
+func newCmdStartedEvent(t time.Time, cmd *exec.Cmd) *event {
+	return newEvent(EventTypeCmdStarted, t, map[string]string{
+		"cmd": cmdString(cmd),
+	})
+}
+
+func newCmdFinishedEvent(t time.Time, cmd *exec.Cmd, startTime time.Time, err error) *event {
+	f := map[string]string{
+		"cmd":      cmdString(cmd),
+		"duration": t.Sub(startTime).String(),
+	}
+	if err != nil {
+		f["err"] = err.Error()
+	}
+	return newEvent(EventTypeCmdFinished, t, f)
+}
+
+func newFinishedEvent(t time.Time, startTime time.Time, err error) *event {
+	f := map[string]string{
+		"duration": t.Sub(startTime).String(),
+	}
+	if err != nil {
+		f["err"] = err.Error()
+	}
+	return newEvent(EventTypeFinished, t, f)
+}
+
+func (e *event) Type() EventType {
+	return e.E
+}
+
+func (e *event) Time() time.Time {
+	return e.T
+}
+
+func (e *event) Fields() map[string]string {
+	fields := make(map[string]string, 0)
+	for key, value := range e.F {
+		fields[key] = value
+	}
+	return fields
+}
+
+func (e *event) MarshalJSON() ([]byte, error) {
+	return json.Marshal(map[string]interface{}{
+		"type":   e.E.String(),
+		"time":   e.T,
+		"fields": e.F,
+	})
+}
+
+type runnerOptions struct {
+	FastFail          bool
+	MaxConcurrentCmds int
+	EventHandler      EventHandler
+	Timer             Timer
+}
+
+func newRunnerOptions() *runnerOptions {
+	return &runnerOptions{
+		defaultFastFail,
+		defaultMaxConcurrentCmds,
+		defaultEventHandler,
+		defaultTimer,
+	}
+}
+
+type runner struct {
+	options *runnerOptions
+}
+
+func newRunner(options ...RunnerOption) *runner {
+	runnerOptions := newRunnerOptions()
+	for _, option := range options {
+		option(runnerOptions)
+	}
+	return &runner{runnerOptions}
+}
+
+func (r *runner) Run(cmds []*exec.Cmd) error {
+	// do not want to acquire lock in the signal handler
+	// do there is a race condition where err could be set to
+	// errCmdFailed or not set at all even after an interrupt happens
+	var err error
+	doneC := make(chan struct{})
+	cmdControllers := make([]*cmdController, len(cmds))
+	for i, cmd := range cmds {
+		cmdControllers[i] = newCmdController(cmd, r.options.EventHandler, r.options.Timer)
+	}
+
+	signalC := make(chan os.Signal, 1)
+	signal.Notify(signalC, os.Interrupt)
+	go func() {
+		for _ = range signalC {
+			err = errInterrupted
+			doneC <- struct{}{}
```

Comment:
That should be ok, because we will effectively exit on the first value that is received on `doneC`.